### DISCUSSION TITLE
AE/AGC Algoritms

### DISCUSSION THREAD

**Discussion by dgalland** [raspberrypi/picamera2#175] [Q&A]
These two algorithms are grouped in the documentation.
What is the expected effect of the AeEnable flag?
What is the meaning of AeLocked ?
I noticed that AeLocked changes to True long before the algorithm has converged and moreover the convergence is really long.
See the trace in an another discussion
https://github.com/raspberrypi/picamera2/discussions/169
I see also in the documentation a speed parameter of the AGC algorithm, is it possible to access it with Picamera2 ?

**Comment by davidplowman** [comment]
Hi, disabling the AeEnable flag should stop the exposure and analogue gain from being adjusted automatically, of course it takes a little while for this to happen.

As you have probably read elsewhere, there is some work going on to do with digital gain and I'll return to look at AeLocked after that, the implementation of it will certainly be changing in any case.

The AEC/AGC speed parameter cannot be accessed through libcamera, though I believe you can change it in the tuning file.

**Reply by dgalland** [reply]
Thank you David for your answer.
The operation of the AE algorithm is strange.
Let's say we start with a fixed exposure of 6257 and we set AeEnable=True the trace is below
- The convergence is really long, exposure varies slowly
- AeLocked is set too early
- At the end only when the right exposure 3142 is obtained AGC adjusts the digital gain -> OK
I didn't see how to adjust the speed parameter in the tuning file

AeLocked: False Exposure: 6257 Analog: 1.0 Digital: 1.0
AeLocked: False Exposure: 6257 Analog: 1.0 Digital: 1.0
AeLocked: False Exposure: 6257 Analog: 1.0 Digital: 1.0
AeLocked: False Exposure: 6257 Analog: 1.0 Digital: 1.0
AeLocked: False Exposure: 5685 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 5228 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 4857 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 4571 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 4342 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 4142 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3971 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3828 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3571 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3428 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3314 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3257 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3228 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3200 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3171 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3171 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3171 Analog: 1.0 Digital: 1.0
AeLocked: True Exposure: 3142 Analog: 1.0 Digital: 1.0057817697525024
AeLocked: True Exposure: 3142 Analog: 1.0 Digital: 1.0054798126220703
AeLocked: True Exposure: 3142 Analog: 1.0 Digital: 1.0053212642669678
AeLocked: True Exposure: 3142 Analog: 1.0 Digital: 1.0052447319030762
AeLocked: True Exposure: 3142 Analog: 1.0 Digital: 1.0052207708358765
AeLocked: True Exposure: 3142 Analog: 1.0 Digital: 1.0052177906036377
AeLocked: True Exposure: 3142 Analog: 1.0 Digital: 1.005207896232605

**Comment by davidplowman** [comment]
Thanks for the example. I agree it seems that AeLocked gets set inappropriately. I will get round to looking at this, but I need to sort out the digital gain issues first. There's a bit of under-the-hood work to do first in how it deals with some of the delays, and we need to be agreeing that with all of the parties involved in the project.

There are possibly some issues to think about with AeLocked. How locked is locked? Is it enough to be "reasonably close" so that a picture looks OK? Or do we mean "totally 100% converged"? Maybe there are folks with different requirements and so it might need a parameter in the tuning file?

As regards the "speed" question, I think you can set (for example) `"speed": 0.4,` in the JSON file under the AGC section (the default value appears to be 0.2).
